core/control-plane/contract/src/main/java/org/eclipse/dataspaceconnector/contract/offer/ContractOfferServiceImpl.java

```diff
@@ -55,15 +57,21 @@ public ContractOfferServiceImpl(ParticipantAgentService agentService, ContractDe
     @NotNull
     public Stream<ContractOffer> queryContractOffers(ContractOfferQuery query, Range range) {
         var agent = agentService.createFor(query.getClaimToken());
-
-        return definitionService.definitionsFor(agent, range)
+        var offers = definitionService.definitionsFor(agent)
                 .flatMap(definition -> {
-                    var assets = assetIndex.queryAssets(definition.getSelectorExpression());
+                    var spec = QuerySpec.Builder.newInstance().filter(definition.getSelectorExpression().getCriteria())
+                            .build();
                     return Optional.of(definition.getContractPolicyId())
                             .map(policyStore::findById)
-                            .map(policy -> assets.map(asset -> createContractOffer(definition, policy.getPolicy(), asset)))
+                            .map(policy -> assetIndex.queryAssets(spec).map(asset -> createContractOffer(definition, policy.getPolicy(), asset)))
                             .orElseGet(Stream::empty);
-                });
+                })
+                // todo: change this: take limit/offset from query
+                .skip(range.getFrom())
+                .limit(range.getTo() - range.getFrom())
+                .collect(Collectors.toList());
+        return offers.stream();
```

Comment:
I would avoid to collect and stream again

Reply:
yeah, i had that for debugging. should be removed.